### PR TITLE
[CLI-693][WB-3962][WB-3827][WB-3651][CLI-578] Console redirect fixes

### LINKED ISSUE #1873: pudb ui when using wandb
**Describe the bug**

I use pudb to debug my code. Currently, when I use wandb, it will affect the pudb UI.
Usually pudb should fill the whole terminal window, whereas now it becomes fixed sized and very small.

Also reported here: https://github.com/CompVis/taming-transformers/blob/master/main.py#L441

Any clue why that happens?
**To Reproduce**

**Expected behavior**
Using wandb should not affect pudb?

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**

### PR DESCRIPTION
**Pull request by farizrahman4u**
Fixes #1873

Epic:
https://wandb.atlassian.net/browse/WB-4317

Bugs:
https://wandb.atlassian.net/browse/WB-3962
https://wandb.atlassian.net/browse/WB-3827
https://wandb.atlassian.net/browse/WB-3651
https://wandb.atlassian.net/browse/CLI-578

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -15,3 +15,4 @@ protobuf>=3.12.0
 PyYAML
 typing; python_version < '3.5'
 enum34; python_version < '3.4'
+git+https://github.com/selectel/pyte#egg=pyte
```
Comment (raubitsj) [inline requirements.txt:18]: I don't think this is a good idea, Lets chat offline about it.

**Review by raubitsj — CHANGES_REQUESTED** [review]
Regression fails to run on this branch.  All runs fail with:
```
Traceback (most recent call last):
  File "mnist_cnn.py", line 15, in <module>
    import wandb
  File "/home/jeff/.pyenv/versions/v368/lib/python3.6/site-packages/wandb/__init__.py", line 32, in <module>
    from wandb.errors.term import termsetup, termlog, termerror, termwarn
  File "/home/jeff/.pyenv/versions/v368/lib/python3.6/site-packages/wandb/errors/term.py", line 2, in <module>
    import click
ModuleNotFoundError: No module named 'click'
```

File: tests/test_redir.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        o = CapList()
+        r = cls("stdout", [o.append])
+        model = tf.keras.models.Sequential()
+        model.add(tf.keras.layers.Dense(10, input_dim=10))
+        model.compile(loss="mse", optimizer="sgd")
+        r.install()
+        epochs = 5
+        model.fit(np.zeros((10000, 10)), np.ones((10000, 10)), epochs=epochs)
+        r.uninstall()
+        assert len(o) in (epochs * 2, epochs * 2 + 1)  # Allow 1 offs
```
Comment (vanpelt) [inline tests/test_redir.py:118]: Curious what the "1 off" comes from?

File: wandb/sdk/lib/redirect.py [hunk 3]
```diff
@@ -1,301 +1,624 @@
...
+                if i is not None:  # python 3 w/ unbuffered i/o: we need to keep writing
+                    while i < len(data):
+                        i += self._orig_src.write(data[i:])
+                self.src_wrapped_stream.flush()
+            except OSError:
+                return
+            try:
+                self._emulator.write(data.decode("utf-8"))
+            except UnicodeDecodeError:
+                pass  # TODO(frz): partial unicode character?
```
Comment (vanpelt) [inline wandb/sdk/lib/redirect.py:624]: This is a BEAST.  It's looking good, let's add some doc strings to describe what these various classes and methods are doing.

File: tests/test_redir.py (hunk 2 above)
Comment (farizrahman4u) [reply tests/test_redir.py:118]: Windows vs linux

**Comment by raubitsj** [comment]
We are ignoring the codecov/patch warnings for this PR as console is very difficult to cover in the circle environment.

File: wandb/sdk/wandb_run.py [hunk 4]
```diff
@@ -2039,6 +2057,35 @@ def close(self) -> None:
             self.lock.release()
 
 
+class CRDedupedFile(WriteSerializingFile):
```
Comment (raubitsj) [inline wandb/sdk/wandb_run.py:2060]: CRDedupedFile seems out of place. can you find somewhere else to put this?
maybe in lib/something
along with WriteSerializingFile maybe?

**Review by raubitsj — COMMENTED** [review]
Can you pull master into this branch also.

**Review by raubitsj — APPROVED** [review]
Wow, lots of great stuff in here.  merging now.

**Pull request merged**